pkg/storage/storage.go

```diff
@@ -0,0 +1,403 @@
+}
+
+func (s *storage) PatchPost(post *sgtmpb.Post) error {
+	return s.db.Omit(clause.Associations).Create(&post).Error
+}
+
+func (s *storage) GetNumberOfPostsByKind() ([]*sgtmpb.PostByKind, error) {
+	var postsByKind []*sgtmpb.PostByKind
+	err := s.db.
+		Model(&sgtmpb.Post{}).
+		// Where(sgtmpb.Post{Visibility: sgtmpb.Visibility_Public}).
+		Select(`kind, count(*) as quantity`).
+		Group("kind").
+		Find(&postsByKind).
+		Error
+	if err != nil {
+		return nil, err
+	}
+	return postsByKind, nil
+}
+
+func (s *storage) GetTotalDuration() (int64, error) {
+	var totalDuration int64
+	err := s.db.
+		Model(&sgtmpb.Post{}).
+		Select("sum(duration) as total_duration").
+		Where(sgtmpb.Post{
+			Kind: sgtmpb.Post_TrackKind,
+			//Visibility: sgtmpb.Visibility_Public,
+		}).
+		First(&totalDuration).
+		Error
+	if err != nil {
+		return 0, err
+	}
+	return totalDuration, nil
+}
+
+func (s *storage) GetPostBySugID(postSlug string) (*sgtmpb.Post, error) {
+	query := s.db.
+		Preload("Author").
+		Preload("RelationshipsAsSource").
+		Preload("RelationshipsAsSource.TargetPost").
+		Preload("RelationshipsAsSource.TargetUser").
+		Preload("RelationshipsAsTarget").
+		Preload("RelationshipsAsTarget.SourcePost").
+		Preload("RelationshipsAsTarget.SourceUser")
+	id, err := strconv.ParseInt(postSlug, 10, 64)
+	if err == nil {
+		query = query.Where(sgtmpb.Post{ID: id, Kind: sgtmpb.Post_TrackKind})
+	} else {
+		query = query.Where(sgtmpb.Post{Slug: postSlug, Kind: sgtmpb.Post_TrackKind})
+	}
+	var post *sgtmpb.Post
+	err = query.First(&post).Error
+	if err != nil {
+		return nil, err
+	}
+	return post, nil
+}
+
+func (s *storage) GetPostComments(postID int64) ([]*sgtmpb.Post, error) {
+	var postComments []*sgtmpb.Post
+	err := s.db.
+		Where(sgtmpb.Post{
+			Kind:         sgtmpb.Post_CommentKind,
+			TargetPostID: postID,
+			Visibility:   sgtmpb.Visibility_Public,
+		}).
+		Preload("Author").
+		Find(&postComments).
+		Error
+	if err != nil {
+		return nil, err
+	}
+	return postComments, nil
+}
+
+func (s *storage) GetUserBySlug(slug string) (*sgtmpb.User, error) {
+	var user *sgtmpb.User
+	err := s.db.
+		Where("LOWER(slug) = ?", slug).
+		First(&user).
+		Error
+	if err != nil {
+		return nil, nil
+	}
+	return user, nil
+}
+
+func (s *storage) GetCalendarHeatMap(authorID int64) ([]int64, error) {
+	var timestamps []int64
+	err := s.db.Model(&sgtmpb.Post{}).
+		Select(`sort_date/1000000000 as timestamp`).
+		Where(sgtmpb.Post{
+			AuthorID:   authorID,
+			Kind:       sgtmpb.Post_TrackKind,
+			Visibility: sgtmpb.Visibility_Public,
+		}).
+		Pluck("timestamp", &timestamps).
+		Error
+	if err != nil {
+		return nil, err
+	}
+	return timestamps, nil
+}
+
+func (s *storage) UpdatePost(post *sgtmpb.Post) error {
+	return s.db.Omit(clause.Associations).Where("id = ?", post.ID).Save(post).Error
+}
+
+func (s *storage) GenericUpdatePost(model interface{}, fields interface{}) error {
+	return s.db.Omit(clause.Associations).Model(model).Updates(fields).Error
+}
```

Comment:
```suggestion
func (s *storage) UpdatePost(post *sgtmpb.Post, updates interface{}) error {
	return s.db.Omit(clause.Associations).Model(post).Updates(updates).Error
}
```

I suggest you to keep only this one with is the best one in all cases